File: voiceroom/src/main/java/cn/rongcloud/voiceroom/room/dialogFragment/CreatorSettingFragment.java
package cn.rongcloud.voiceroom.room.dialogFragment;


import android.view.View;

import androidx.appcompat.widget.AppCompatTextView;
import androidx.constraintlayout.widget.Guideline;

import com.rongcloud.common.utils.ImageLoaderUtil;

import cn.rong.combusis.common.base.BaseBottomSheetDialogFragment;
import cn.rong.combusis.common.ui.dialog.ConfirmDialog;
import cn.rong.combusis.music.MusicManager;
import cn.rong.combusis.provider.user.User;
import cn.rong.combusis.sdk.event.wrapper.EToast;
import cn.rongcloud.voiceroom.R;
import cn.rongcloud.voiceroom.room.VoiceRoomModel;
import cn.rongcloud.voiceroom.ui.uimodel.UiSeatModel;
import de.hdodenhof.circleimageview.CircleImageView;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.functions.Action;
import io.reactivex.rxjava3.functions.Consumer;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;

/**
 * 房主点击自己的头像弹窗
 */
public class CreatorSettingFragment extends BaseBottomSheetDialogFragment implements View.OnClickListener {


    private Guideline glBg;
    private CircleImageView ivMemberPortrait;
    private AppCompatTextView tvMemberName;
    private AppCompatTextView btnOutOfSeat;
    private AppCompatTextView btnMuteSelf;
    private ConfirmDialog confirmDialog;
    private VoiceRoomModel voiceRoomModel;
    private UiSeatModel uiSeatModel;
    private User user;

    public CreatorSettingFragment(VoiceRoomModel voiceRoomModel, UiSeatModel uiSeatModel, User user) {
        super(R.layout.fragmeng_new_creator_setting);
        this.voiceRoomModel = voiceRoomModel;
        this.uiSeatModel = uiSeatModel;
        this.user = user;
    }

    @Override
    public void initView() {
        glBg = (Guideline) getView().findViewById(R.id.gl_bg);
        ivMemberPortrait = (CircleImageView) getView().findViewById(R.id.iv_member_portrait);
        tvMemberName = (AppCompatTextView) getView().findViewById(R.id.tv_member_name);
        btnOutOfSeat = (AppCompatTextView) getView().findViewById(R.id.btn_out_of_seat);
        btnMuteSelf = (AppCompatTextView) getView().findViewById(R.id.btn_mute_self);
        btnMuteSelf.setText(uiSeatModel.isMute() ? "打开麦克风" : "关闭麦克风");
        tvMemberName.setText(user.getUserName());
        ImageLoaderUtil.INSTANCE.loadImage(getContext(), ivMemberPortrait, user.getPortraitUrl(), R.drawable.default_portrait);
    }

    @Override
    public void initListener() {
        btnOutOfSeat.setOnClickListener(this::onClick);
        btnMuteSelf.setOnClickListener(this::onClick);
        super.initListener();
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.btn_out_of_seat) {//是否下麦围观
            //判断是否在播放音乐
            if (voiceRoomModel.isPlayingMusic()) {
                showMusicPauseTip();
            } else {
                leaveSeat();
            }
        } else if (v.getId() == R.id.btn_mute_self) {
            //操作关闭还是打开麦克风
            voiceRoomModel.creatorMuteSelf()
                    .subscribe(new Action() {
                        @Override
                        public void run() throws Throwable {
                            dismiss();
                        }
                    }, new Consumer<Throwable>() {
                        @Override
                        public void accept(Throwable throwable) throws Throwable {
                            EToast.showToast(throwable.getMessage());
                        }
                    });
        }
    }

    /**
     * 离开麦位
     */
    private void leaveSeat() {
        voiceRoomModel.leaveSeat()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Action() {
                    @Override
                    public void run() throws Throwable {
                        dismiss();
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Throwable {
                        EToast.showToast(throwable.getMessage());
                    }
                });
    }

    private void showMusicPauseTip() {
        confirmDialog = new ConfirmDialog(requireContext(), "播放音乐中下麦会导致音乐中断，是否确定下麦?",
                true, "确定", "取消", null, new Function0<Unit>() {
            @Override
            public Unit invoke() {
                //确定
                MusicManager.get().stopPlayMusic();
                leaveSeat();
                return null;
            }
        });
        confirmDialog.show();
    }
}

File: app/src/main/java/cn/rongcloud/voiceroomdemo/mvp/fragment/voiceroom/creatorsetting/CreatorSettingFragment.kt
/*
 * Copyright © 2021 RongCloud. All rights reserved.
 */

package cn.rongcloud.voiceroomdemo.mvp.fragment.voiceroom.creatorsetting

import cn.rongcloud.annotation.HiltBinding
import cn.rongcloud.voiceroomdemo.R
import com.rongcloud.common.base.BaseBottomSheetDialogFragment
import cn.rongcloud.mvoiceroom.net.bean.respond.VoiceRoomBean
import com.rongcloud.common.extension.loadPortrait
import com.rongcloud.common.extension.ui
import com.rongcloud.common.utils.AccountStore
import dagger.hilt.android.AndroidEntryPoint
import com.rongcloud.common.ui.dialog.ConfirmDialog
import kotlinx.android.synthetic.main.fragmeng_creator_setting.*
import javax.inject.Inject

/**
 * @author gusd
 * @Date 2021/06/28
 */
@HiltBinding(value = ICreatorView::class)
@AndroidEntryPoint
class CreatorSettingFragment(view: ICreatorView, private val roomInfoBean: VoiceRoomBean) :
    BaseBottomSheetDialogFragment(R.layout.fragmeng_creator_setting),
    ICreatorView by view {

    @Inject
    lateinit var presenter: CreatorSettingPresenter


    fun showMusicPauseTip(confirmBlock: () -> Unit) {
        ConfirmDialog(
            requireContext(),
            "播放音乐中下麦会导致音乐中断，是否确定下麦?",
            true,
        ) {
            confirmBlock()
        }.apply {
            show()
        }
    }

    override fun initView() {
        btn_out_of_seat.setOnClickListener {
            if (presenter.isPlayingMusic()) {
                showMusicPauseTip {
                    presenter.stopPlayMusic()
                    presenter.leaveSeat()
                }
            } else {
                presenter.leaveSeat()
            }
        }
        btn_mute_self.setOnClickListener {
            presenter.muteMic()
        }
        tv_member_name.text = roomInfoBean.createUser?.userName

        iv_member_portrait.loadPortrait(AccountStore.getUserPortrait())
    }

    override fun fragmentDismiss() {
        super.fragmentDismiss()
        dismiss()
    }

    override fun onMuteChange(isMute: Boolean) {
        super.onMuteChange(isMute)
        ui {
            if (isMute) {
                btn_mute_self.text = "打开麦克风"
            } else {
                btn_mute_self.text = "关闭麦克风"
            }
        }
    }
}